experimental/uefi/runtime/src/lib.rs

```diff
@@ -52,24 +52,36 @@ impl<T: Channel> Framed<T> {
     }
 
     pub fn read_frame(&mut self) -> anyhow::Result<Frame> {
-        let mut length_buf = [0; FRAME_LENGTH_ENCODED_SIZE];
-        self.inner.recv(&mut length_buf)?;
-        let length = FrameLength::from_be_bytes(length_buf);
-        let mut body: Vec<u8> = vec![0; length.into()];
-        self.inner.recv(&mut body)?;
+        let length = {
+            let mut length_bytes = [0; FRAME_BYTES_SIZE];
+            self.inner.recv(&mut length_bytes)?;
+            FrameLength::from_le_bytes(length_bytes)
+        };
+
+        let body = {
+            let body_length: u32 = length - FRAME_BYTES_SIZE as u32;
+            let mut body: Vec<u8> = vec![
+                0;
+                usize::try_from(body_length).expect(
+                    "the supported pointer size is smaller than the frame length"
+                )
+            ];
+            self.inner.recv(&mut body)?;
+            body
+        };
+
         Ok(Frame { body })
     }
 
     pub fn write_frame(&mut self, frame: Frame) -> anyhow::Result<()> {
-        let length: FrameLength = FrameLength::try_from(frame.body.len())
-            .map_err(anyhow::Error::msg)
-            .context("the frame body is too large")?;
-        let encoded_length = length.to_be_bytes();
-        let mut encoded_frame: Vec<u8> =
-            Vec::with_capacity(encoded_length.len() + frame.body.len());
-        encoded_frame.extend(encoded_length);
-        encoded_frame.extend(frame.body);
-        self.inner.send(&encoded_frame)?;
+        let length_bytes = {
+            let length = FrameLength::try_from(FRAME_BYTES_SIZE + frame.body.len())
+                .map_err(anyhow::Error::msg)
+                .context("the frame is too large")?;
+            length.to_be_bytes()
+        };
+        self.inner.send(&length_bytes)?;
+        self.inner.send(&frame.body)?;
```

Comment:
Calling send in two increments avoids the extra allocation that sometimes occurred in the previous approach using extend